frontend/src/main/scala/bloop/bsp/BloopBspServices.scala

```diff
@@ -421,6 +424,84 @@ final class BloopBspServices(
     Task.eval { originToCompileStores.remove(params.originId); () }.executeAsync
   }
 
+  def linkProjects(
+      userProjects: Seq[ProjectMapping],
+      state: State,
+      compileArgs: List[String],
+      originId: Option[String],
+      logger: BspServerLogger
+  ): BspResult[bsp.CompileResult] = {
+    import bloop.engine.tasks.LinkTask.{linkJS, linkMainWithNative}
+
+    def doLink(
+        project: Project,
+        newState: State
+    ) =
+      project.platform match {
+        case platform @ Js(config, _, _) =>
+          val cmd = Commands.Link(List(project.name))
+          val targetDir = ScalaJsToolchain.linkTargetFrom(project, config)
+          linkJS(cmd, project, newState, false, None, targetDir, platform).map { linkState =>
+            val result = if (linkState.status == ExitStatus.LinkingError) {
+              Right(bsp.CompileResult(originId, bsp.StatusCode.Error, None, None))
+            } else {
+              Right(bsp.CompileResult(originId, bsp.StatusCode.Ok, None, None))
+            }
+            (linkState, result)
+          }
+        case Jvm(_, _, _, _, _, _) => ???
+        case platform @ Native(config, _, userMainClass) =>
+          userMainClass match {
+            case None =>
+              logger.error("No main class defined in project")
+              Task.now(
+                (newState, Right(bsp.CompileResult(originId, bsp.StatusCode.Error, None, None)))
+              )
+            case Some(value) =>
+              val cmd = Commands.Link(List(project.name))
+              val target = ScalaNativeToolchain.linkTargetFrom(project, config)
+              linkMainWithNative(cmd, project, newState, value, target, platform).map { linkState =>
+                val result = if (linkState.status == ExitStatus.LinkingError) {
+                  Right(bsp.CompileResult(originId, bsp.StatusCode.Error, None, None))
+                } else {
+                  Right(bsp.CompileResult(originId, bsp.StatusCode.Ok, None, None))
+                }
+                (linkState, result)
+              }
+          }
+      }
+
+    compileProjects(userProjects, state, compileArgs, originId, logger).flatMap {
+      case (newState, Right(CompileResult(_, StatusCode.Ok, _, _))) =>
+        val linkExecution: Task[Seq[(State, Either[Response.Error, CompileResult])]] =
+          Task.sequence(userProjects.map { case (_, p) => doLink(p, newState) })
+
+        linkExecution.materialize.map {
+          case Success(linkRunsSeq) =>
+            linkRunsSeq
+              .map(_._2)
+              .foldLeft(Option.empty[CompileResult])((_, res) => // TODO propogate stuff better here
```

Comment:
This logic is somewhat lifted from the test function, but it doesn't really work that well here. I'm not entirely sure what the best strategy is for combining the compileresults (actually linker results)